File: src/backend/repository/JsonRepository.go
package repository

import (
    "log"
	"io/ioutil"
    "encoding/json"
)

// Repository type Json, and it has functions of IRepository.
type JsonRepository struct {
	areas AreaJson
	emotes EmoteJson
	textMap map[uint]SightseeingLogTextJson
	sLogs []SightseeingLogJson
	weathers WeatherJson
	weatherChances WeatherChanceJson
}

// Get sightseeing logs from json
func (r *JsonRepository) LoadSightseeingLogs() ([]SightseeingLog, error) {
	var result []SightseeingLog
	for _, slog := range r.sLogs {
		item := SightseeingLog {
			ItemNo: slog.ItemNo,
			AreaKey: slog.AreaKey,
			AreaName: r.areas[slog.AreaKey].JP,
			CoordinateX: slog.CoordinateX,
			CoordinateY: slog.CoordinateY,
			StartHour: slog.StartHour,
			EndHour: slog.EndHour,
			EmoteId: slog.EmoteID,
			EmoteName: r.emotes[slog.EmoteID].JP,
			Weather1Key: slog.Weather1Key,
			Weather1Name: r.weathers[slog.Weather1Key].JP,
			Weather2Key: slog.Weather2Key,
			Description: r.textMap[slog.ItemNo].Description.JP,
		}

		if slog.Weather2Key != nil {
			str := r.weathers[*slog.Weather2Key].JP
            item.Weather2Name = &str
        }

		result = append(result, item)
	}
    return result, nil
}

// Get weather chances from json
func (r *JsonRepository) LoadWeatherChances() ([]WeatherChance, error) {
	var result []WeatherChance
    for area, chances := range r.weatherChances {
        for i, chance := range chances {
            result = append(result, WeatherChance {
                AreaKey:     area,
                WeatherKey:  chance.WeatherKey,
                Chance:      chance.Chance,
                ChanceIndex: i + 1,
            })
        }
    }

    return result, nil
}

// Read json file and convert it to type T struct
func GetJson[T any](fileName string) T {
	var result T
	
	data, err := ioutil.ReadFile("data/" + fileName)
    if err != nil {
        log.Fatalf("Error reading JSON file: %v", err)
    }

	err = json.Unmarshal(data, &result)
	if err != nil {
        log.Fatalf("Error deseliallize JSON file: %v", err)
	}

	return result
}

// Create JsonRepository instance
func NewJsonRepository() *JsonRepository {
    // Read json and keep static datas
	areas := GetJson[AreaJson]("areas.json")
	emotes := GetJson[EmoteJson]("emotes.json")
	texts := GetJson[[]SightseeingLogTextJson]("sightseeingLogTexts.json")
	sLogs := GetJson[[]SightseeingLogJson]("sightseeingLogs.json")
	weathers := GetJson[WeatherJson]("weathers.json")
	weatherChances := GetJson[WeatherChanceJson]("weatherChances.json")
	
	textMap := make(map[uint]SightseeingLogTextJson)
	for _, item := range texts {
		textMap[item.ItemNo] = item
	}

    return &JsonRepository {
		areas: areas,
		emotes: emotes,
		textMap: textMap,
		sLogs: sLogs,
		weathers: weathers,
		weatherChances: weatherChances,
	}
}



// ==============
//  for json parse
type LocalizedText struct {
    JP string `json:"jp"`
    EN string `json:"en"`
}

type AreaJson map[string]LocalizedText

type EmoteJson map[uint]LocalizedText

type SightseeingLogTextJson struct {
    ItemNo       uint           `json:"item_no"`
    Description  *LocalizedText `json:"description"`
    Hint         *LocalizedText `json:"hint"`
}

type SightseeingLogJson struct {
    ItemNo        uint     `json:"item_no"`
    AreaKey       string   `json:"area_key"`
    CoordinateX   float64  `json:"coordinate_x"`
    CoordinateY   float64  `json:"coordinate_y"`
	StartHour     uint     `json:"in_game_start_hour"`
    EndHour       uint     `json:"in_game_end_hour"`
    EmoteID       uint     `json:"emote_id"`
    Weather1Key   string   `json:"weather1_key"`
    Weather2Key   *string  `json:"weather2_key"`
}

type WeatherJson map[string]LocalizedText

type WeatherChanceJsonData struct {
    WeatherKey string `json:"weather_key"`
    Chance     int    `json:"chance"`
}

type WeatherChanceJson map[string][]WeatherChanceJsonData

File: src/features/repository/jsonRepository.ts
import { AreasJson, EmotesJson, SightseeingLogsJson, SightseeingLogTextsJson, WeatherChancesJson, WeathersJson } from "../interface/jsonClass"
import { SightseeingLog, WeatherChance } from "../interface/dataClass";
import { IRepository } from "../interface/interface"
import path from 'path';
import fs from 'fs';

class JsonRepository implements IRepository {

    private completesLoading = false;
    private areas: AreasJson = {};
    private emotes: EmotesJson = {};
    private logs: SightseeingLogsJson = [];
    private texts: SightseeingLogTextsJson = [];
    private chances: WeatherChancesJson = {};
    private weathers: WeathersJson = {};
    private static instance: JsonRepository;

    private constructor() {}
  
    // Make this class a singleton to use cache of static json data.
    static getInstance(): JsonRepository {
        if (!JsonRepository.instance) {
            JsonRepository.instance = new JsonRepository();
            
            // Load jsons. This may take a long, so run it async.
            JsonRepository.instance.loadJsons();
        }
        return JsonRepository.instance
    }

    private async loadJsons() {
        console.log("[JsonRepository] Json loading started.");
        console.time("[JsonRepository] Json loading completed. time");
        this.areas = this.loadFile("areas.json");
        this.emotes = this.loadFile("emotes.json");
        this.logs = this.loadFile("sightseeingLogs.json");
        this.texts = this.loadFile("sightseeingLogTexts.json");
        this.chances = this.loadFile("weatherChances.json");
        this.weathers = this.loadFile("weathers.json");
        this.completesLoading = true;
        console.timeEnd("[JsonRepository] Json loading completed. time");
    }

    // Load sightseeing logs data
    LoadSightseeingLogs(): SightseeingLog[] {
        if (!this.completesLoading) {
            return [];
        }

        return this.logs.map(log => {
            const area = this.areas[log.area_key].jp;
            const emote = this.emotes[log.emote_key].jp;
            const weather1 = this.weathers[log.weather1_key].jp;
            const weather2 = log.weather2_key != null ? this.weathers[log.weather2_key].jp : undefined;
            const textJson = this.texts.find(f => f.item_no == log.item_no);
            const text = textJson ? textJson.description.jp : "";
            const hint = textJson ? textJson.hint.jp : "";

            return {
                ItemNo: log.item_no,
                AreaKey: log.area_key,
                AreaName: area,
                CoordinateX: log.coordinate_x,
                CoordinateY: log.coordinate_y,
                StartHour: log.in_game_start_hour,
                EndHour: log.in_game_end_hour,
                EmoteKey: log.emote_key,
                EmoteName: emote,
                Weather1Key: log.weather1_key,
                Weather1Name: weather1,
                Weather2Key: log.weather2_key,
                Weather2Name: weather2,
                Description: text,
                Hint: hint,
            }
        })
    }

    // Load weather chances
    LoadWeatherChances(): Map<string, WeatherChance[]> {
        const weatherChanceMap = new Map<string, WeatherChance[]>();
        if (!this.completesLoading) {
            return weatherChanceMap;
        }

        for (const area in this.chances) {
            const chances = this.chances[area].map(entry => ({
                WeatherKey: entry.weather_key,
                Chance: entry.chance
            }));
            weatherChanceMap.set(area, chances);
        }
    
        return weatherChanceMap;
    }
    
    private loadFile(fileName: string) {
        const filePath = path.join(process.cwd(), 'public', 'data', fileName);
        const fileContents = fs.readFileSync(filePath, 'utf8');
        return JSON.parse(fileContents);
    }
}

export default JsonRepository.getInstance();